components/telemetry-controller/controllers/loggingconfiguration_controller.go

```diff
@@ -0,0 +1,318 @@
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *LoggingConfigurationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	var config telemetryv1alpha1.LoggingConfiguration
+	if err := r.Get(ctx, req.NamespacedName, &config); err != nil {
+		log.Error(err, "unable to fetch LoggingConfiguration")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	updatedCm, err := r.syncConfigMap(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync ConfigMap")
+		return ctrl.Result{}, err
+	}
+
+	updatedFiles, err := r.syncFiles(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync mounted files")
+		return ctrl.Result{}, err
+	}
+
+	updatedEnv, err := r.syncEnvironment(ctx, &config)
+	if err != nil {
+		log.Error(err, "failed to sync environment variables")
+		return ctrl.Result{}, err
+	}
+
+	if updatedCm || updatedFiles || updatedEnv {
+		if err := r.Update(ctx, &config); err != nil {
+			log.Error(err, "cannot update LoggingConfiguration")
+			return ctrl.Result{}, err
+		}
+
+		if err := r.deleteFluentBitPods(ctx, log); err != nil {
+			log.Error(err, "cannot delete fluent bit pods")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *LoggingConfigurationReconciler) getOrCreateConfigMap(ctx context.Context, name types.NamespacedName) (corev1.ConfigMap, error) {
+	var cm corev1.ConfigMap
+	if err := r.Get(ctx, name, &cm); err != nil {
+		if errors.IsNotFound(err) {
+			cm = corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      name.Name,
+					Namespace: name.Namespace,
+				},
+			}
+			if err := r.Create(ctx, &cm); err != nil {
+				return cm, err
+			}
+		} else {
+			return cm, err
+		}
+	}
+	return cm, nil
+}
+
+func (r *LoggingConfigurationReconciler) syncConfigMap(ctx context.Context, config *telemetryv1alpha1.LoggingConfiguration) (bool, error) {
+	cm, err := r.getOrCreateConfigMap(ctx, r.FluentBitConfigMap)
+	if err != nil {
+		return false, err
+	}
+	cmKey := config.Name + ".conf"
+
+	// Add or remove Fluent Bit configuration sections
+	if config.DeletionTimestamp != nil {
+		if cm.Data != nil && controllerutil.ContainsFinalizer(config, configMapFinalizer) {
+			delete(cm.Data, cmKey)
+			controllerutil.RemoveFinalizer(config, configMapFinalizer)
+		}
+	} else {
+		fluentBitConfig := generateFluentBitConfig(config.Spec.Sections)
+		if cm.Data == nil {
+			data := make(map[string]string)
+			data[cmKey] = fluentBitConfig
+			cm.Data = data
+		} else {
+			if oldConfig, hasKey := cm.Data[cmKey]; hasKey && oldConfig == fluentBitConfig {
+				// Nothing changed
+				return false, nil
+			}
+			cm.Data[cmKey] = fluentBitConfig
+		}
+		controllerutil.AddFinalizer(config, configMapFinalizer)
```

Comment:
In the example they guard both an addition and removal of a finalizer with a check: https://book.kubebuilder.io/reference/using-finalizers.html. Not sure if it's needed, just wanted to mention ;)